lib/Optimizer/GraphOptimizer/NodeSplitting.cpp

```diff
@@ -0,0 +1,1561 @@
+    for (const auto &outputIdxMap : outputIdxAndMaps) {
+      auto outputCheckedRange = outputIdxMap.second(splitOutputSlice);
+      outputRanges[outputIdxMap.first] = outputCheckedRange.second;
+    }
+
+    // Create input Slice nodes.
+    for (const auto &inputIdxMap : inputIdxAndMaps) {
+      auto inputIdx = inputIdxMap.first;
+      auto &inputRange = inputRanges[inputIdxMap.first];
+      Type outTy = Type::newShape(*(node->getNthInput(inputIdx).getType()),
+                                  inputRange.getSizes());
+      auto nodeName =
+          clone->getName().str() + ".SliceInput" + std::to_string(inputIdx);
+      auto *inputSlice = F->createSlice(nodeName, node->getNthInput(inputIdx),
+                                        inputRange.getStarts(), &outTy);
+      clone->setNthInput(inputIdx, inputSlice);
+    }
+
+    // Set clone split output type. The original node output type is not
+    // modified because the clone owns its output types.
+    TypeRef splitOutputType = F->getParent()->uniqueTypeWithNewShape(
+        node->getType(splitOutputIdx), splitOutputSlice.getSizes());
+    clone->getNthResult(splitOutputIdx).setTypeUnsafe(splitOutputType);
+
+    // Set clone output types. The original node output types are not
+    // modified because the clone owns its output types.
+    for (const auto &outputIdxMap : outputIdxAndMaps) {
+      auto &outputRange = outputRanges[outputIdxMap.first];
+      TypeRef outputType = F->getParent()->uniqueTypeWithNewShape(
+          node->getType(outputIdxMap.first), outputRange.getSizes());
+      clone->getNthResult(outputIdxMap.first).setTypeUnsafe(outputType);
+    }
+
+    // Modify clone.
+    splitNodeModifier(node, clone, inputRanges, outputRanges);
+
+    // Verify clone.
+    RETURN_ERR_IF_NOT(clone->verify(),
+                      "Invalid node obtained during node splitting!");
+
+    // Add clone to function.
+    F->addNode(clone);
+
+    // Add clone to vector.
+    splitNodes[sliceIdx] = clone;
+
+    // Merge the partial outputs of this clone.
+    for (size_t outIdx = 0, outIdxEnd = node->getNumResults();
+         outIdx < outIdxEnd; outIdx++) {
+      auto nodeName =
+          clone->getName().str() + ".MergeOutput" + std::to_string(outIdx);
+      mergedOutputs[outIdx] = F->createInsertTensor(
+          nodeName, mergedOutputs[outIdx], clone->getNthResult(outIdx),
+          outputRanges[outIdx].getStarts());
+    }
+  }
+
+  // Replace all the node outputs with the merged outputs.
+  for (size_t outIdx = 0, outIdxEnd = node->getNumResults(); outIdx < outIdxEnd;
+       outIdx++) {
+    node->getNthResult(outIdx).replaceAllUsesOfWith(mergedOutputs[outIdx]);
+  }
+
+  return splitNodes;
+}
+
+///===---------------------------------------------------------------------===//
+///                                    Conv
+///===---------------------------------------------------------------------===//
+static std::tuple<bool, DimRange, DimPads>
```

Comment:
I'd suggest defining a struct with these fields and semantically meaningful fields names and return it here. It will make the code easier to understand.

Reply:
Done.